external_apps/bid/bid_action/create_bid.go

```diff
@@ -0,0 +1,273 @@
+	"strconv"
+	"time"
+
+	"github.com/pkg/errors"
+	"github.com/tendermint/tendermint/libs/kv"
+
+	"github.com/Oneledger/protocol/action"
+	"github.com/Oneledger/protocol/data/keys"
+)
+
+var _ action.Msg = &CreateBid{}
+
+type CreateBid struct {
+	BidConvId  bid_data.BidConvId    `json:"bidConvId"`
+	AssetOwner keys.Address          `json:"assetOwner"`
+	AssetName  string                `json:"assetName"`
+	AssetType  bid_data.BidAssetType `json:"assetType"`
+	Bidder     keys.Address          `json:"bidder"`
+	Amount     action.Amount         `json:"amount"`
+	Deadline   int64                 `json:"deadline"`
+}
+
+var _ action.Tx = &CreateBidTx{}
+
+type CreateBidTx struct {
+}
+
+func (c CreateBidTx) Validate(ctx *action.Context, signedTx action.SignedTx) (bool, error) {
+	createBid := CreateBid{}
+	err := createBid.Unmarshal(signedTx.Data)
+	if err != nil {
+		return false, errors.Wrap(action.ErrWrongTxType, err.Error())
+	}
+
+	//validate basic signature
+	err = action.ValidateBasic(signedTx.RawBytes(), createBid.Signers(), signedTx.Signatures)
+	if err != nil {
+		return false, err
+	}
+	err = action.ValidateFee(ctx.FeePool.GetOpt(), signedTx.Fee)
+	if err != nil {
+		return false, err
+	}
+
+	// the currency should be OLT
+	currency, ok := ctx.Currencies.GetCurrencyById(0)
+	if !ok {
+		panic("no default currency available in the network")
+	}
+	if currency.Name != createBid.Amount.Currency {
+		return false, errors.Wrap(action.ErrInvalidAmount, createBid.Amount.String())
+	}
+
+	//Check if bid ID is valid(if provided)
+	if len(createBid.BidConvId) > 0 && createBid.BidConvId.Err() != nil {
+		return false, bid_data.ErrInvalidBidConvId
+	}
+
+	//Check if bidder and owner address is valid oneLedger address(if bid id is not provided)
+	if len(createBid.BidConvId) == 0 {
+		err = createBid.Bidder.Err()
+		if err != nil {
+			return false, errors.Wrap(action.ErrInvalidAddress, err.Error())
+		}
+
+		err = createBid.AssetOwner.Err()
+		if err != nil {
+			return false, errors.Wrap(action.ErrInvalidAddress, err.Error())
+		}
+	}
+
+	return true, nil
+}
+
+func (c CreateBidTx) ProcessCheck(ctx *action.Context, tx action.RawTx) (bool, action.Response) {
+	ctx.Logger.Detail("Processing CreateBid Transaction for CheckTx", tx)
+	return runCreateBid(ctx, tx)
+}
+
+func (c CreateBidTx) ProcessDeliver(ctx *action.Context, tx action.RawTx) (bool, action.Response) {
+	ctx.Logger.Detail("Processing CreateBid Transaction for DeliverTx", tx)
+	return runCreateBid(ctx, tx)
+}
+
+func (c CreateBidTx) ProcessFee(ctx *action.Context, signedTx action.SignedTx, start action.Gas, size action.Gas) (bool, action.Response) {
+	return action.BasicFeeHandling(ctx, signedTx, start, size, 1)
+}
+
+func runCreateBid(ctx *action.Context, tx action.RawTx) (bool, action.Response) {
+	// if this is to create bid conversation, everything except bidConvId is needed
+	// if this is just to add an offer from bidder, only needs bidConvId, bidder(to sign), amount
```

Comment:
there are two scenarios:
1. create bid conversation AND bid offer(create both in one tx)
2. add an offer to an existing bid conversation.

the reason for this is that if we separate create bid conversation and create bid offer into two separate tx, we always need to do 2 txs at beginning, because an empty bid conversation without any offers makes no sense